### ISSUE TITLE
MultiprocessIterator deadlocks with cupy==2.0.0a1, works fine with cupy==1.0.0.1

### ISSUE THREAD

**Issue by zori** [chainer/chainer#3075]
I am not sure this is related to https://github.com/chainer/chainer/issues/2909

My code runs on 4 GPUs using the following requirements:
```
chainer==2.0.1
cupy==1.0.1
cycler==0.10.0
Cython==0.25.2
filelock==2.0.11
matplotlib==2.0.2
nose==1.3.7
numpy==1.13.1
olefile==0.44
Pillow==4.1.1
protobuf==3.3.0
pyparsing==2.2.0
python-dateutil==2.6.1
pytz==2017.2
six==1.10.0
tqdm==4.14.0
opencv==3.1.0
```
OS: Ubuntu 14.04.5 LTS
CUDA: 8.0.44
cuDNN: 5.1.5

The above runs fine, but changing the `cupy==1.0.1` -> `cupy==2.0.0a1` results in a deadlock in MultiprocessIterator. After killing the process, I get the following stacktrace
[cupy_deadlock_stacktrace.txt](https://github.com/chainer/chainer/files/1186393/cupy_deadlock_stacktrace.txt)

Sorry, but I don't currently have a minimal example. I'll post that to the cupy issue tracker as well.

**Comment by delta2323** [comment]
Thank you for your report. We will investigate the problem along with #2909.

**Comment by sharpstill** [comment]
@delta2323 I have encounter the same issue in MultiprocessIterator .
The MultiprocessIterator must to check again carefully

**Comment by delta2323** [comment]
@sharpstill Thank you for sharing the information.

@zori Could you try the combination of chainer==v3.0.0b1 and cupy==2.0.0b1?

**Comment by zori** [comment]
@delta2323 works!

I did, however, notice a small performance penalty. chainer==2.0.1 and cupy==1.0.0.1:
```
        95 iter, 0 epoch / 500 epochs
   0.89046 iters/sec. Estimated time to finish: 1 day, 4:56:33.121165.
```
chainer==v3.0.0b1 and cupy==2.0.0b1:
```
        80 iter, 0 epoch / 500 epochs
   0.69135 iters/sec. Estimated time to finish: 1 day, 13:17:02.110467.
```

**Comment by delta2323** [comment]
Thank you for trying it. Also thank you for sharing the performance evaluation. We need to investigate the performance more as another problem.

**Comment by delta2323** [comment]
@sharpstill Could you try as @zori did?

**Comment by sharpstill** [comment]
@delta2323  if I have some problem in one of the examples, what happened after I use chainer.dataset.DatasetMixin and raise IndexError in def get_example? could the sub-process become zombie process? I notice that is zombie process after I raise IndexError

**Comment by delta2323** [comment]
So do you mean the process that `MultiProcessIterator` used gets Zombie when the dataset the iterator handles raises `IndexError` in `get_example`?

**Comment by sharpstill** [comment]
@delta2323 yes,  I use get_example function, and for some reason, My data has some bad example datum which can only know who is bad during running training process. So I must raise IndexError .
But this cause MultiProcessIterator. By the way , I use chainer==v3.0.0b1 and cupy==2.0.0b1.

**Comment by sharpstill** [comment]
@delta2323 Does MultiProcessIterator bug solved?

**Comment by okuta** [comment]
@zori Maybe, This issue was fixed by #3076.
Cloud you check dead lock on your environment with CuPy v3.0.0rc ?

**Comment by beam2d** [comment]
Closed for now. Feel free to reopen if you still have the problem!
